File: Lista_7/lista7_ex4.c
/* Exercício 4:
Suponha que criamos uma estrutura para armazenar dados de pessoas e um vetor para armazenar dados de várias pessoas:
typedef struct Pessoa{
    int rg;
    int cpf;
    char nome[80];
}Pessoa;

Pessoa cadastro[100];

Suponha que o vetor esteja ordenado em ordem crescente por valor de RG. 
Implemente uma função de busca por RG, que opera como a busca binária, e que caso exista uma pessoa no cadastro com o RG a ser buscado, 
devolve o índice deste no cadastro, e devolve -1 caso não exista uma pessoa com o RG a ser buscado.
*/
#include <stdio.h>
#include <stdlib.h>

typedef struct Pessoa{
    int rg;
    int cpf;
    char nome[80];
}Pessoa;

void armazenaDados(Pessoa cadastro[], int n);
void imprimirDados(Pessoa cadastro[], int n);
void ordenaRG(Pessoa cadastro[], int n);
int buscaBinaria(Pessoa cadastro[], int n, int chave);

int main () {
int n, res, chave;
Pessoa cadastro[100];

    printf("Numero de pessoas a serem cadastradas: ");
    scanf("%d", &n);

    armazenaDados(cadastro, n);

    printf("\nPESSOAS CADASTRADAS:\n");
    imprimirDados(cadastro, n);

    printf("\nDigite o numero de RG a ser buscado: ");
    scanf("%d", &chave);

    res = buscaBinaria(cadastro, n, chave);

    if(res == -1) {
        printf("\nRG nao foi encontrado!\n");
    } else {
        printf("\nRG foi encontrado!\n");
    }

    printf("\nPESSOAS CADASTRADAS:\n");
    imprimirDados(cadastro, n);

return 0;
}

void armazenaDados(Pessoa cadastro[], int n) {
    for(int i=0; i < n; i++) {
        printf("\nInsira %d pessoa:\n", i+1);
        getchar();
        printf("Nome: ");
        fgets(cadastro[i].nome, 80, stdin);
        cadastro[i].nome[strlen(cadastro[i].nome)-1] = '\0';
        printf("RG: ");
        scanf("%d", &cadastro[i].rg);
        printf("CPF: ");
        scanf("%d", &cadastro[i].cpf);
    }
}

void imprimirDados(Pessoa cadastro[], int n) {
    for(int i=0; i < n; i++) {
        printf("\nPESSOA %d:", i+1);
        printf("\nNome: %s", cadastro[i].nome);
        printf("\nRG: %d", cadastro[i].rg);
        printf("\nCPF: %d\n", cadastro[i].cpf);
    }
}

void ordenaRG(Pessoa cadastro[], int n) {
    Pessoa temp[1];
    for(int i=0; i < n; i++) {
        for(int j=i+1; j < n; j++) {
            if(cadastro[i].rg > cadastro[j].rg) {
                temp[0] = cadastro[i];
                cadastro[i] = cadastro[j];
                cadastro[j] = temp[0];
            }
        }
    }
}

int buscaBinaria(Pessoa cadastro[], int n, int chave) {
    int ini=0, fim=n-1, meio;

    ordenaRG(cadastro, n);

    while(ini <= fim) {
        meio = (ini+fim)/2;
        if(cadastro[meio].rg == chave) {
            return meio;
        } else if(cadastro[meio].rg > chave) {
            fim = meio-1;
        } else {
            ini = meio+1;
        }
    }
return -1;
}

File: Lista_7/Lista7_ex4.py
class Pessoa:

	def __init__(self):
		self.__rg = 0
		self.__cpf = 0
		self.__nome = ''
	
	def get_rg(self):
		return self.__rg

	def get_cpf(self):
		return self.__cpf

	def get_nome(self):
		return self.__nome

	def cadastra(self):
		self.__rg = int(input("RG: "))
		self.__cpf = int(input("CPF: "))
		self.__nome = input("Nome: ")

class Sistema (Pessoa):

	def __init__(self):
		self.__lista = []
		self.__numero = 0

	def get_numero_pessoas(self):
		self.__numero = int(input("Numero de pessoas: "))
		return self.__numero

	def get_lista(self, pessoa):		
		pessoa.cadastra()
		self.__lista.append(pessoa)
		return self.__lista

	def mostra(self):
		for p in self.__lista:
			print('RG: ' + str(p.get_rg()) +
			      '; CPF: ' + str(p.get_cpf()) +
			      '; Nome: ' + p.get_nome())	

	def busca_binaria(self, chave):
		ini = 0
		fim = self.__numero - 1
		ordena = sorted(self.__lista, key=lambda obj: obj.get_rg())
		
		while ini <= fim:
			meio = int((ini + fim) / 2)
			if ordena[meio].get_rg() == chave:
				return meio
			elif ordena[meio].get_rg() > chave:
				fim = meio - 1
			else:
				ini = meio + 1
		return -1
		
if __name__ == "__main__":

	s = Sistema()
	q = 's'

	print("### Cadastro de Pessoas ###\n")
	n = s.get_numero_pessoas()
	for i in range(n):
		print("----------")		
		p = Pessoa()
		s.get_lista(p)
	
	while(q == 's'):
		chave = int(input("\nDigite o numero de RG a ser buscado: "))
		r = s.busca_binaria(chave)
		if r == -1:
			print("\nRG nao foi encontrado!\n")
		else:
			print("\nRG foi encontrado!\n")

		q = input("\nFazer nova busca (s/n): ")
		if q == 'n':
			break

	print("\n### Pessoas cadastradas ###\n")
	s.mostra()
